lib/matplotlib/axes/_axes.py

```diff
@@ -5638,24 +5639,59 @@ def _pcolorargs(funcname, *args, allmatch=False):
             raise TypeError(
                 'Incompatible X, Y inputs to %s; see help(%s)' % (
                 funcname, funcname))
-        if allmatch:
-            if (Nx, Ny) != (ncols, nrows):
+
+        if shading == 'auto':
+            if ncols == Nx and nrows == Ny:
+                shading = 'nearest'
+            else:
+                shading = 'flat'
+
+        if shading == 'flat':
+            if not (ncols in (Nx, Nx - 1) and nrows in (Ny, Ny - 1)):
                 raise TypeError('Dimensions of C %s are incompatible with'
                                 ' X (%d) and/or Y (%d); see help(%s)' % (
                                     C.shape, Nx, Ny, funcname))
-        else:
-            if not (ncols in (Nx, Nx - 1) and nrows in (Ny, Ny - 1)):
+            if (ncols == Nx and nrows == Ny):
+                cbook.warn_deprecated("3.3",
+                    message="shading='flat' when X and Y have the same "
+                            "dimensions as C is deprecated since %(since)s.  "
+                            "Either specify the corners of the quadrilaterals "
+                            "with X and Y, or pass shading='auto', 'nearest' "
+                            "or 'gouraud', or set rcParams['pcolor.shading']")
+                C = C[:Ny - 1, :Nx - 1]
```

Comment:
It looks like you are missing the mixed cases like `C.shape == (Ny+1, Nx)`. Did we handle them before?

Reply:
No we didn't handle that, and I think thats OK for simplicity's sake. If someone has a more complicated grid I feel they can wrangle it themselves....

Reply:
Actually, we did handle it, because in the "flat" case we always did the indexing (your line 5661), which "worked" with the mixed case.  In the code you have here, the mixed case will fail somewhere downstream instead of being caught with an exception, or simply handled as it was originally.

Reply:
I think you could fix it by changing "and" to "or" in your line 5654, and then out-denting line 5661.

Reply:
Actually, 5654 could also change "and" to "or".  Warn if *either* dimension isn't right.

Reply:
Yep that now emits the warning for:
`ax.pcolormesh(x, y[:-1], C[:-1, :-1], vmin=0, vmax=1, shading='flat')`